src/frontend/src/flows/authenticate/postMessageInterface.ts

```diff
@@ -21,14 +23,67 @@ export interface AuthResponseSuccess {
   userPublicKey: Uint8Array;
 }
 
-export interface AuthResponseFailure {
-  kind: "authorize-client-failure";
-  text: string;
+/**
+ * All information required to process an authentication request received from
+ * a client application.
+ */
+export interface AuthContext {
+  /**
+   * Information sent by the client application.
+   */
+  authRequest: AuthRequest;
+  /**
+   * Origin of the message.
+   */
+  requestOrigin: string;
+  /**
+   * Callback to send a result back to the sender. We currently only send
+   * either a success message or nothing at all.
+   */
+  postMessageCallback: (message: AuthResponseSuccess) => void;
 }
 
-export type AuthResponse = AuthResponseSuccess | AuthResponseFailure;
-
 // A message to signal that the II is ready to receive authorization requests.
 export const READY_MESSAGE = {
   kind: "authorize-ready",
 };
+
+/**
+ * Set up an event listener for window post messages and wait for the authorize
+ * authentication request from the client application.
+ */
+export default async function waitForAuthRequest(): Promise<AuthContext | null> {
+  const result = new Promise<AuthContext>((resolve, reject) => {
+    // Set up an event listener for receiving messages from the client.
+    window.addEventListener("message", async (event) => {
+      const message = event.data;
+      if (message.kind === "authorize-client") {
+        console.log("Handling authorize-client request.");
+        resolve({
+          authRequest: message,
+          requestOrigin: event.origin,
+          postMessageCallback: (responseMessage) =>
+            (event.source as Window).postMessage(responseMessage, event.origin),
+        });
+      } else {
+        console.error(
+          `Message of unknown kind received: ${JSON.stringify(message)}`
+        );
+        reject();
+      }
+    });
+  });
+
+  // Send a message to indicate we're ready.
+  // NOTE: Because `window.opener.origin` cannot be accessed, this message
+  // is sent with "*" as the target origin. This is safe as no sensitive
+  // information is being communicated here.
+  if (window.opener !== null) {
+    window.opener.postMessage(READY_MESSAGE, "*");
+  } else {
+    // If there's no `window.opener` a user has manually navigated to "/#authorize".
+    // Signal that there will never be an authentication request incoming.
+    return Promise.resolve(null);
```

Comment:
Could we start the function with that? i.e. 

``` ts
export default async function waitForAuthRequest(): Promise<AuthContext | null> {
  if(window.opener === null) {
    return Promise.resolve(null);
  }

  ... 
} 
```

at least it'd be clear why we need `| null`, and then you don't have to read the big promise block above to get to that. It's basically "failing early".